beacon-chain/verification/data_column.go

```diff
@@ -0,0 +1,330 @@
+	if err != nil {
+		return errors.Wrap(columnErrBuilder(ErrSlotNotAfterParent), "parent root not in forkchoice")
+	}
+	if parentSlot >= dv.dataColumn.Slot() {
+		return ErrSlotNotAfterParent
+	}
+	return nil
+}
+
+// SidecarDescendsFromFinalized represents the spec verification:
+// [REJECT] The current finalized_checkpoint is an ancestor of the sidecar's block
+// -- i.e. get_checkpoint_block(store, block_header.parent_root, store.finalized_checkpoint.epoch) == store.finalized_checkpoint.root.
+func (dv *RODataColumnVerifier) SidecarDescendsFromFinalized() (err error) {
+	defer dv.recordResult(RequireSidecarDescendsFromFinalized, &err)
+	if !dv.fc.HasNode(dv.dataColumn.ParentRoot()) {
+		log.WithFields(logging.DataColumnFields(dv.dataColumn)).Debug("parent root not in forkchoice")
+		return columnErrBuilder(ErrSidecarNotFinalizedDescendent)
+	}
+	return nil
+}
+
+// SidecarInclusionProven represents the spec verification:
+// [REJECT] The sidecar's kzg_commitments field inclusion proof is valid as verified by verify_data_column_sidecar_inclusion_proof(sidecar).
+func (dv *RODataColumnVerifier) SidecarInclusionProven() (err error) {
+	defer dv.recordResult(RequireSidecarInclusionProven, &err)
+	if err = blocks.VerifyKZGInclusionProofColumn(dv.dataColumn); err != nil {
+		log.WithError(err).WithFields(logging.DataColumnFields(dv.dataColumn)).Debug("sidecar inclusion proof verification failed")
+		return columnErrBuilder(ErrSidecarInclusionProofInvalid)
+	}
+	return nil
+}
+
+// SidecarKzgProofVerified represents the spec verification:
+// [REJECT] The sidecar's column data is valid as verified by verify_data_column_sidecar_kzg_proofs(sidecar).
+func (dv *RODataColumnVerifier) SidecarKzgProofVerified() (err error) {
+	defer dv.recordResult(RequireSidecarKzgProofVerified, &err)
+	ok, err := dv.verifyDataColumnCommitment(dv.dataColumn)
+	if err != nil {
+		log.WithError(err).WithFields(logging.DataColumnFields(dv.dataColumn)).Debug("kzg commitment proof verification failed")
+		return columnErrBuilder(ErrSidecarKzgProofInvalid)
+	}
+	if !ok {
+		log.WithFields(logging.DataColumnFields(dv.dataColumn)).Debug("kzg commitment proof verification failed")
+		return columnErrBuilder(ErrSidecarKzgProofInvalid)
+	}
+	return nil
+}
+
+// SidecarProposerExpected represents the spec verification:
+// [REJECT] The sidecar is proposed by the expected proposer_index for the block's slot
+// in the context of the current shuffling (defined by block_header.parent_root/block_header.slot).
+// If the proposer_index cannot immediately be verified against the expected shuffling, the sidecar MAY be queued
+// for later processing while proposers for the block's branch are calculated -- in such a case do not REJECT, instead IGNORE this message.
+func (dv *RODataColumnVerifier) SidecarProposerExpected(ctx context.Context) (err error) {
+	defer dv.recordResult(RequireSidecarProposerExpected, &err)
+	e := slots.ToEpoch(dv.dataColumn.Slot())
+	if e > 0 {
+		e = e - 1
```

Comment:
This specific case is not tested. (I guess it's important to test it since it is the majority common case.)